more changes (#17)
infinispan/Infinispan.Hotrod.Core: README.md

# Infinispan Hot Rod .NET Core Demo

This branch provides a demonstration of the Hot Rod .NET Core client with remote caches on an Infinispan Server instance.
### Goal

> This demo aims to show:
> - the current maturity level of the new .NET Core client library;
> - how to include it in your application;
> - how to store and query entries in remote cache;
> - how the client handles Protobuf serialization.

Reviewer: ```suggestion
This demo aims to show you the current maturity level of the Hot Rod .NET Core client as well as how to:

- Include ithe client in your application.
- Store and query entries in a remote cache.
- Handle Protobuf serialization of C# data types.
```

### Some highlights

> - Concurrency: the Hot Rod client API is asynchronous by default and concurrency is supported. To show that, the demo runs all PUT commands asynchronously ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L111-L122)).

Reviewer: ```suggestion
- Concurrency: The Hot Rod client API is asynchronous by default and supports concurrency. To show these capabilites the demo runs all PUT commands asynchronously ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L111-L122)).
```

> - Simplicity: the client is shipped as a .nuget package and can be easily imported ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L7)).

Reviewer: ```suggestion
- Simplicity: The client is shipped as a `.nuget` package that you can easily import ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L7)).
```

> - Protobuf support: using the default Protobuf/Grpc tools ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L8-L9)), the C# data types hierarchy can be generated and used in the application.

Reviewer: ```suggestion
- Protobuf support: Using the default Protobuf/GRPC tools ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Query.csproj#L8-L9)), the C# data types hierarchy can be generated and used in the application.
```

### Infinispan Query

#### Server setup
To run queries the Infinispan server needs a `.proto` description of the queriable data types. This demo takes care of uploading the `.proto` file onto the server. ([link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/e2efac6591741d23ff92c6253bf1257a60ea8879/demo/Query/Program.cs#L69))

#### Client Marshaller

> Infinispan uses `application/x-protostream` media type to serialize remote query. [Protostream](https://github.com/infinispan/protostream) is a Protobuf based protocol that allows multiple types of object to be serialized onto the same data stream. Simplifying a little bit: client and server share a map that assigns an integer number to each queriable object type; this is usually done defining the `TypeId` in the `.proto` file. (see [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/app.proto#L5) and [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/review.proto#L5)). This typeId is sent in front of each Protobuf message sent on the wire.

Reviewer: ```suggestion
Infinispan uses the `application/x-protostream` media type to serialize data for remote queries. [ProtoStream](https://github.com/infinispan/protostream) is a Protobuf-based protocol that allows multiple object types to be serialized in the same data stream. Simplifying a little bit, the client and server share a map that assigns an integer number to each query-able object type; this is usually done defining the `TypeId` in the `.proto` schema file. (see [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/app.proto#L5) and [link](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/a648993db9cd97ebff2186a6f3f5ef64b37517da/demo/Query/Protos/review.proto#L5)). This `TypeId` is sent in front of each Protobuf message transmitted across the wire.
```

On the client side the application must provide a Protostream marshaller for its own data types. That could sound scary, but Protobuf can do most of the hard work for this.

> A protostream essentially must:

Reviewer: ```suggestion
A ProtoStream marshaller essentially does the following:
```

> - set the correct `typeId` in the protostream message when marshalling data to the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L17-L24));

Reviewer: ```suggestion
- Sets the correct `TypeId` in the ProtoStream message when marshalling data to the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L17-L24));
```

> - use the correct Protobuf parser when unmarshalling from the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L31-L37)).

Reviewer: ```suggestion
- Uses the correct Protobuf parser when unmarshalling objects from the server ([see](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Marshaller.cs#L31-L37)).
```

Once the marshaller has been implemented, the plain C# classes generated by the Protobuf framework can be used in the application development.

> Maybe it worth to highlight that, using protostream, the data format is independent from the specific programming language; this means that data store in the cache can be accessed by C#, Java, C++ clients. Actually the demo shows a round trip interoperability between different data representations:

Reviewer: ```suggestion
It is worth highlighting that, using ProtoStream, the data format is independent from the specific programming language. This means that data stored in the cache can be accessed by C#, Java, C++ clients. Actually the demo shows round-trip interoperability between different data representations:
```

> - [json->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L117) marshalling;

Reviewer: ```suggestion
- [JSON->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L117) marshalling.
```

> - [C#->protostream](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L118) marshalling;

Reviewer: ```suggestion
- [C#->ProtoStream](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L118) marshalling.
```

> - [protostream->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L46-L51) unmarshalling;

Reviewer: ```suggestion
- [ProtoStream->C#](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L46-L51) unmarshalling.
```

> - [C#->json](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L52) marshalling.

Reviewer: ```suggestion
- [C#->JSON](https://github.com/infinispan/Infinispan.Hotrod.Core/blob/234362df176512f23d0eaef171a26b6f5ccf9489/Program.cs#L52) marshalling.
```

### Conclusion

> With the Infinispan Hot Rod .NET Core client it's easy for .NET applications to store and query remote caches, in an language independent format, while working with native C# data types.

Reviewer: ```suggestion
With the Infinispan Hot Rod .NET Core client it's easy for .NET applications to store and query remote caches, in an language neutral format, while working with native C# data types.
```
